snappy/helpers.go

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (C) 2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+/* This module gathers a set of helper functions and structures shared
+   by the various commands in snapweb
+*/
+
+package snappy
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"net/http"
+	"os"
+	"os/signal"
+	"path/filepath"
+	"strings"
+	"sync"
+	"syscall"
+
+	"github.com/snapcore/snapd/client"
+)
+
+// IsDeviceManaged determines if the device is in the 'managed' state
+func IsDeviceManaged() bool {
+	client := client.New(nil)
+
+	sysInfo, err := client.SysInfo()
+	if err != nil {
+		return false
+	}
+
+	if sysInfo.OnClassic || sysInfo.Managed {
+		return true
+	}
+
+	return false
+}
+
+func unixDialer(socketPath string) func(string, string) (net.Conn, error) {
+	file, err := os.OpenFile(socketPath, os.O_RDWR, 0666)
+	if err == nil {
+		file.Close()
+	}
+
+	return func(_, _ string) (net.Conn, error) {
+		return net.Dial("unix", socketPath)
+	}
+}
+
+// MakePassthroughHandler maps a snapd API to a snapweb HTTP handler
+func MakePassthroughHandler(socketPath string, prefix string) http.HandlerFunc {
```

Comment:
This is a general mechanism used in snapweb, and re-used as such in webconf, to map an HTTP call to the equivalent snapd API call. Instead of having a dedicated handler for each API in snapweb, and then linking the snapd client code to generate the unix call, Michael proposed to use this passthru, and minimize code duplication.
Note that In snapweb, this is only accessible if the access control token is sent along with the request.